[e2e] Refactor RBAC tests to use client-go instead of raw HTTP

demonCoder95 opened the pull request:
## Background
The current RBAC e2e test implementation breaks with stricter CA Cert validation enforcement when doing TLS handshake with the api-server in EKS clusters.

This rewrites the RBAC test cases to use client-go, in order to avoid handling raw HTTP calls, and also creates a cleaner and more readable Gingko suite for RBAC e2e tests. This change for now, simply retains the existing tests (in `authorisation_test.go`) and writes these in addition to those. We can drop the older ones later.

## Summary of Changes
1. Drop `PodSecurityPolicy` related e2e tests as PSPs have been removed now.
2. Use `Context` and `When` containers for better organization of our RBAC e2e specs. The current implementation puts all the test cases in a single spec.
3. The utility functions are moved in `authorization_utils.go` file to keep the main `authorization.go` file cleaner.
4. Certain permission functionality from the Authz webhook has been moved to the admission-controller and this test suite being based on `SubjectAccessReview` can't verify that. These tests will be moved to the admission-controller e2e tests and removed from here.

I decided not to generalise the attribute expansion logic and make the code too brittle and complicated. The code for expansion right now is very verbose, but it's clearer to follow, understand and debug.

demonCoder95 committed:
add the rbac test suite description

demonCoder95 committed:
add rbac e2e utils

mikkeloscar commented on test/e2e/authorization_utils.go:
```diff
@@ -0,0 +1,38 @@
+	return cs.AuthorizationV1().SubjectAccessReviews().Create(ctx, &sar, metav1.CreateOptions{})
+}
+
+func createLocalClient() (kubernetes.Interface, error) {
```
The e2e framework already provides a way to get a client (see other e2e tests in this repo). Doing it this way would make it more complicated to run locally as you would need to setup a hardcoded path: `/workdir/test/e2e/kubeconfig`

demonCoder95 replied:
I'll check. Thanks!

demonCoder95 replied:
Turns out that's much easier than this 😆   Just need to do
```golang
f := framework.NewDefaultFramework("framework-name")
cs := f.ClientSet
```

Done!

demonCoder95 committed:
setup the testing interface between test cases and utils backend

demonCoder95 committed:
implement logic to evaluate output of a testcase

demonCoder95 committed:
add logic to expand into resourceAttributes and nonResourceAttributes from testCaseData

demonCoder95 committed:
enable role-sync-controller

demonCoder95 committed:
add test cases for ReadOnly group

demonCoder95 commented on test/e2e/authorization_utils.go:
```diff
@@ -0,0 +1,278 @@
+	// expand the testcase data to generate a list of ResourceAttributes
+	resourceAttributes := t.expandResourceAttributes()
+
+	// expand the testcase data to generate a list of NonResourceAttributes
+	// nonResourceAttributes := t.expandNonResourceAttributes()
+
+	// expand the testcase data to generate a list of SubjectAccessReview objects
+	// based on the ResourceAttributes and NonResourceAttributes
+	for _, ra := range resourceAttributes {
+		for _, user := range t.data.users {
+			for _, group := range t.data.groups {
+				sar := authv1.SubjectAccessReview{
+					Spec: authv1.SubjectAccessReviewSpec{
+						ResourceAttributes: &ra,
+						User:               user,
+						Groups:             group,
+					},
+				}
+				sars = append(sars, sar)
+			}
+		}
+	}
+	return sars
+}
+
```
This is a temporary fix to get the tests running, will be refactored into expansions.

demonCoder95 replied:
This is now done.

demonCoder95 committed:
resolve TODO about handling subresources

demonCoder95 committed:
fix bug in resource attribute expansion

demonCoder95 committed:
add test cases for PowerUser, Manual and Emergency groups

demonCoder95 committed:
Refactor how the testCase output is evaluated

We now match each SubjectAccessReview object with the expected result.
We then populate a list of all those that didn't meet expectation.
If a test fails, we use the String() function on the testCaseOutput
to print all these violating SARs.
If there are not such SARs, we consider the test passed.

demonCoder95 committed:
add all expansions in SAR generation

demonCoder95 committed:
implement all the testcase specs

demonCoder95 committed:
improve pretty printing by filtering out empty values

demonCoder95 committed:
cleanup pretty printing, add a comment

demonCoder95 committed:
fix test cases

demonCoder95 commented:
This is now ready for review 🎉

demonCoder95 committed:
remove write e2e tests that are to be covered by admission-controller e2e

linki commented on cluster/config-defaults.yaml:
```diff
@@ -1194,4 +1194,8 @@ teapot_admission_controller_scheduling_controls_enabled: "false"
 
 # role-sync-controller configs
+{{ if eq .Cluster.Environment "e2e" }}
+role_sync_controller_enabled: "true"
```
This would enable the sync controller for non-EKS clusters. Do we want that?

demonCoder95 replied:
Hmm I could change it to EKS only?

My plan is to ultimately keep the `role-sync-controller` enabled and disable the authz webhook at some point, for all clusters. This allows us:
1. single authz setup with no parity between eks/non-eks
2. test the setup by rolling it out on channels and fix issues if we find any

We can hold off on that while we still have some work pending on the admission-controller, I guess? What do you suggest?

demonCoder95 replied:
I updated to only enable this for EKS.

linki replied:
I'm fine with that plan.

But with this, it would create additional bindings in the namespaces that we don't need because auth-webhook is still running. I'm just concerned about confusion (it should be fine from a permissions perspective).

linki commented on test/e2e/authorization.go:
```diff
@@ -0,0 +1,572 @@
+					"persistentvolumeclaims",
+					"configmaps",
+				}
+				tc.data.verbs = []string{"create", "update", "delete", "patch"}
```
We could extract these "static" lists as vars similar to `allGroups` in the top and then re-use them. For example:

```go
	var (
		namespacedResource = []string{
			"pods",
			"apps/deployments",
			"apps/statefulsets",
			"apps/deployments/scale",
			"apps/statefulsets/scale",
			"services",
			"persistentvolumes",
			"persistentvolumeclaims",
			"configmaps",
		}
		globalResources = []string{
			"namespaces",
			"nodes",
			"rbac.authorization.k8s.io/clusterroles",
			"storage.k8s.io/storageclasses",
			"apiextensions.k8s.io/customresourcedefinitions",
		}
		readOperations  = []string{"get", "list", "watch"}
		writeOperations = []string{"create", "update", "delete", "patch"}
	)
```

demonCoder95 replied:
That's actually a good suggestions. I'll do that. Thanks.

demonCoder95 replied:
Added in this commit: https://github.com/zalando-incubator/kubernetes-on-aws/pull/8536/commits/1b13defef78ff9c53dbdc6c43a9e14659f868991

Looks much better 🙂

linki commented on test/e2e/authorization.go:
```diff
@@ -0,0 +1,572 @@
+			})
+			// These should be covered by the admission-controller tests.
+			// They're written here for completeness.
+			g.It("should deny write access in kube-system and visibility namespaces", func() {})
```
Can we implement them but `skip` them? This way we can validate them if we later manage to implement them with RBAC.

demonCoder95 replied:
Sounds good. I'll do it.

demonCoder95 replied:
Done in this commit: https://github.com/zalando-incubator/kubernetes-on-aws/pull/8536/commits/6f7f8c96c5ef8339c41832b8ffdc86a8a9b46b97

I found another write protection test that failed so I skipped that as well, and marked it with a comment in the code.

linki commented on test/e2e/authorization.go:
```diff
@@ -0,0 +1,572 @@
+			// These should be covered by the admission-controller tests.
+			// They're written here for completeness.
+			g.It("should deny write access in kube-system and visibility namespaces", func() {})
+			g.It("should allow write access in namespaces other than kube-system and visibility", func() {})
```
Similar to `"should allow write access in namespaces other than kube-system"` (line 314) we can add the implemetation here 🤷

demonCoder95 replied:
I agree. I'll add.

demonCoder95 replied:
Added in the commit: https://github.com/zalando-incubator/kubernetes-on-aws/pull/8536/commits/6f7f8c96c5ef8339c41832b8ffdc86a8a9b46b97

linki commented on test/e2e/authorization.go:
```diff
@@ -0,0 +1,572 @@
+			})
+			// TODO: Need to verify in the original tests if this is a permission on
+			// the controller-manager or the daemonset-controller.
+			// g.It("should allow to create Pods", func() {})
```
Is this clarified?

demonCoder95 replied:
No 🙁  Do you know what's the right permission here? The current setup uses `daemonset-controller` SA. I'm not sure if we keep it that way or change it to `controller-manager`.

linki replied:
I don't know for sure. Maybe you can `git blame` it to get more context.

linki commented on test/e2e/authorization.go:
```diff
@@ -0,0 +1,572 @@
+			// g.It("should allow to create Pods", func() {})
+		})
+
+		g.When("the default service account is in default namespace", func() {
```
```suggestion
		g.When("the service account is the default service account in the default namespace", func() {
```

demonCoder95 replied:
This was taken care of when these two test cases were combined in: https://github.com/zalando-incubator/kubernetes-on-aws/pull/8536/commits/4eba59b8cb63f1d56b7ef05fad86022bfe76ebae

linki commented on test/e2e/authorization.go:
```diff
@@ -0,0 +1,572 @@
+			g.BeforeEach(func() {
+				tc.data.users = []string{"system:serviceaccount:default:default"}
+			})
+			g.It("should deny to list StatefulSets", func() {
```
Interesting

demonCoder95 replied:
Not sure why we added this check 😀

linki commented on test/e2e/authorization.go:
```diff
@@ -0,0 +1,572 @@
+				tc.data.resources = []string{"apps/statefulsets"}
+				tc.data.verbs = []string{"list"}
+				tc.run(context.TODO(), cs, false)
+				gomega.Expect(tc.output.passed).To(gomega.BeTrue(), tc.output.String())
```
This is the same behaviour as the previous test, right? Can we combine them?

demonCoder95 replied:
I'll check if we can combine. Maybe define the "list stateful sets" as the container block 🤔

linki replied:
Or just one test with this?

```go
tc.data.users = []string{"system:serviceaccount:default:default", "system:serviceaccount:non-default:default"}
```

demonCoder95 replied:
Ah yes. you're right. They're identical 😅 I was focussing too much on copying the tests 1 to 1. I merged them into one test case specifying both users.

Combined them in one test in: https://github.com/zalando-incubator/kubernetes-on-aws/pull/8536/commits/4eba59b8cb63f1d56b7ef05fad86022bfe76ebae

linki commented on test/e2e/authorization.go:
```diff
@@ -0,0 +1,572 @@
+					tc.data.verbs = []string{"update"}
+					// Technically, this should result in access undecided because we allow
+					// access to 'skipper-default-filters' ConfigMap only and we haven't
+					// specified a resource name in the test case.
```
Can we just specify another resource name (like `foo`)?

demonCoder95 replied:
I can do that.

demonCoder95 replied:
I actually tried specifying `foo-configmap` but it's still undecided. Not sure why this is happening. Should I still add the name?

linki commented on test/e2e/authorization.go:
```diff
@@ -0,0 +1,572 @@
+			g.BeforeEach(func() {
+				tc.data.users = []string{"zalando-iam:zalando:service:k8sapi_credentials-provider"}
+			})
+			g.It("should allow to get Secrets in kube-system namespace", func() {
```
(Allow-listing) writing secrets is again done in `admission-controller` 😠 Hopefully we find a better solution in the future.

linki replied:
Actually RBAC also gives credentials provider read/write to secret. Maybe it makes sense to add a test for it as well?

demonCoder95 replied:
Yeah 🤞

demonCoder95 replied:
> Actually RBAC also gives credentials provider read/write to secret. Maybe it makes sense to add a test for it as well?

Looking at[ the `ClusterRoleBinding`](https://github.com/zalando-incubator/kubernetes-on-aws/blob/dev/cluster/manifests/roles/credentials-provider-rbac.yaml#L20-L27) attached to this user, it's allowed all operation on `Secrets` except `delete`. 

I created two test cases, one for deletion, another for "all other" operations: https://github.com/zalando-incubator/kubernetes-on-aws/pull/8536/commits/39e6e18d57eb00c4da9ef0a7d4a1fa4c044c91ee

linki commented on test/e2e/authorization.go:
```diff
@@ -0,0 +1,572 @@
+				g.It("should allow read access", func() {
+					tc.data.verbs = []string{"get", "list", "watch"}
+					tc.run(context.TODO(), cs, true)
+					gomega.Expect(tc.output.passed).To(gomega.BeTrue(), tc.output.String())
```
Let's also add a test for write access?

demonCoder95 replied:
I'll add. Thanks.

demonCoder95 replied:
Added it here: https://github.com/zalando-incubator/kubernetes-on-aws/pull/8536/commits/04827d63f832859c141450ad58b07440dd42b616

linki commented on test/e2e/authorization.go:
```diff
@@ -0,0 +1,572 @@
+					tc.data.verbs = []string{"create", "update", "delete", "patch"}
+					tc.run(context.TODO(), cs, true)
+					gomega.Expect(tc.output.passed).To(gomega.BeTrue(), tc.output.String())
+				})
```
Since we expand the `verbs` anyway, we could also combine this into one test with `"should allow full access"`. Not sure if that makes it clearer but just a style fyi.

demonCoder95 replied:
Actually that sounds a lot better. I can do that.

demonCoder95 replied:
Added here: https://github.com/zalando-incubator/kubernetes-on-aws/pull/8536/commits/1646c0671183e2aa08fea97d3fb23933872c86d5

demonCoder95 committed:
enable role-sync-controller by default only on EKS

demonCoder95 commented:
The remaining comments will be addressed in a follow-up PR, as agreed.

demonCoder95 commented:
👍

linki commented:
👍
